LAH15/20170116/SEG.md

```diff
@@ -0,0 +1,16 @@
+# SEG
+## Nguyễn Đức Huy
+
+## Problem Statement
+Trên trục số người ta sơn n đoạn thẳng, đoạn thẳng thứ i có 2 đầu mút ai và bi. Một số đoạn liên tiếp có thể chồng lên nhau một phần hoặc nằm hẳn bên trong nhau. Cuối cùng, sẽ có một số đoạn được sơn rời nhau. Hãy xác định số lượng các đoạn thẳng như thế và chiều dài của đoạn dài nhất.
+
+## Required Skills
+- Thuật toán sắp xếp.
+- Tư tưởng duyệt từ trái sang phải.
+
+## Step-by-Step Solution
+Đầu tiên, ta thấy rằng nếu mỗi một đầu mút được định nghĩa là một sự kiện, một biến điều khiển là cur=0, với việc khi bắt đầu đoạn sơn thì sẽ cur++ và khi kết thúc sơn thì sẽ là cur--, và sắp xếp lại các thứ tự các sự kiện theo chiều tăng dần từ trái qua phải, và sự kiện cur++ được xử lí trước thì một đoạn được sơn liên tiếp kết thúc khi sau một sự kiện nào đó mà giá trị cur==0. Ta lưu lại biến để xử lý trong lúc for. Độ phức tạp O(nlogn)
```

Comment:
+Đầu tiên, ta thấy rằng nếu mỗi một đầu mút được định nghĩa là một sự kiện, tạo một biến điều khiển cur=0, cur biểu thị cho số đoạn thẳng có đi qua điểm hiện tại. Mỗi lần đi qua 1 điểm có sự kiện là bắt đầu 1 đoạn sơn thì sẽ cur++, và mỗi lần đi qua 1 điểm có sự kiện kết thúc 1 đoạn sơn thì sẽ cur--. Sắp xếp 2 loại sự kiện bắt đầu và kết thúc vào 2 mảng với thứ tự tăng dần từ trái qua phải, xử lý sự kiện cur++ trước thì 1 đoạn sơn liên tiếp được kết thúc khi (cur==0). Sử dụng thêm biến để đếm và tính độ dài trong lúc for. Độ phức tạp O(n log n).